### ISSUE TITLE
Enabling TF to build LLVM AMDGPU backend for XLA

### ISSUE THREAD

**Issue by adityaatluri** [tensorflow/tensorflow#10437]
Hi,
When I enable XLA during configure stage and run `bazel build --config=opt //tensorflow/tools/pip_package:build_pip_package`, it starts pulling [llvm](https://github.com/tensorflow/tensorflow/blob/master/third_party/llvm/llvm.BUILD) and build for ARM, PPC, X86 and NVPTX backends. Is there a way to enable AMDGPU backend? (Matter of fact, all backends supported by LLVM?)

**Comment by petecoup** [comment]
Not sure what will be different for a GPU target, but for a CPU target, I modified the llvm.BUILD to add the target to the llvm_targets = [] list, added the target { "name":, ... "tbl_outs": [(...)] } entries to llvm_target_list, and "cc_library" entries like the other CPU targets (<target>_asm_printer, <target>_code_gen, <target>_desc, <target>_disassembler, <target>_info).  Once I had all of this, in compiler/xla/service/cpu/BUILD I added @llvm//:<target>_code_gen and @llvm//:<target>_disassembler to the deps list for the "cpu_compiler" cc_library.
From what I see, you'd have to do something like this to enable any new target, and there isn't a way to enable "all" of the targets that are available in upstream llvm...and I'm not sure what would change for the AMDGPU GPU target.

**Comment by adityaatluri** [comment]
Hi,
Thank you for response. I tried to add the entries you mentioned and running into build errors.
I documented my observations here:
https://groups.google.com/a/tensorflow.org/forum/#!topic/discuss/Diy-mVlcNPQ

**Comment by petecoup** [comment]
Hmm...
...looking more closely at the AMDGPU backend.  Unlike any of the other current supported backends, it looks like the intrinsics you're interested in are in lib/Target/AMDGPU, rather than what is in include/llvm/IR/IntrinsicsAMDGPU.td...and you're the only target that uses "-gen-intrinsic" directly.  I'm guessing that this is problematic.  The regular CMakeLists.txt rule for this is:
tablegen(LLVM AMDGPUGenIntrinsics.inc -gen-tgt-intrinsic)
 Can you try "-gen-tgt-intrinsic" instead?

**Comment by adityaatluri** [comment]
Ah! Thank you for the insight of CMake->Bazel mapping of -gen* flags. Using the idea, I cleaned up llvm.BUILD file and able to build AMDGPU backend. 
Thank you very much!

**Comment by adityaatluri** [comment]
Hi,
When I changed [llvm.BUILD](https://gist.github.com/adityaatluri/8b8ca1f326d6f1537c17c98db84c6ae3) build to work for AMDGPU target, I ran [tf2.py](https://github.com/gpu0/tf-code/blob/master/tf2.py), it is giving out following error:
```
Traceback (most recent call last):
  File "tf1.py", line 2, in <module>
    import tensorflow as tf
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/__init__.py", line 24, in <module>
    from tensorflow.python import *
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/__init__.py", line 49, in <module>
    from tensorflow.python import pywrap_tensorflow
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/pywrap_tensorflow.py", line 52, in <module>
    raise ImportError(msg)
ImportError: Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/pywrap_tensorflow.py", line 41, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 28, in <module>
    _pywrap_tensorflow_internal = swig_import_helper()
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
ImportError: /usr/local/lib/python2.7/dist-packages/tensorflow/python/_pywrap_tensorflow_internal.so: undefined symbol: _ZN4llvm6AMDGPU9isComputeEj


Failed to load the native TensorFlow runtime.

See https://www.tensorflow.org/install/install_sources#common_installation_problems

for some common reasons and solutions.  Include the entire stack trace
above this error message when asking for help.
```
Here are the configure steps I tried:
```
Please specify the location of python. [Default is /usr/bin/python]:
Found possible Python library paths:
  /usr/local/lib/python2.7/dist-packages
  /usr/lib/python2.7/dist-packages
Please input the desired Python library path to use.  Default is [/usr/local/lib/python2.7/dist-packages]

Using python library path: /usr/local/lib/python2.7/dist-packages
Please specify optimization flags to use during compilation when bazel option "--config=opt" is specified [Default is -march=native]:
Do you wish to use jemalloc as the malloc implementation? [Y/n]
jemalloc enabled
Do you wish to build TensorFlow with Google Cloud Platform support? [y/N]
No Google Cloud Platform support will be enabled for TensorFlow
Do you wish to build TensorFlow with Hadoop File System support? [y/N]
No Hadoop File System support will be enabled for TensorFlow
Do you wish to build TensorFlow with the XLA just-in-time compiler (experimental)? [y/N] y
XLA JIT support will be enabled for TensorFlow
Do you wish to build TensorFlow with VERBS support? [y/N]
No VERBS support will be enabled for TensorFlow
Do you wish to build TensorFlow with OpenCL support? [y/N]
No OpenCL support will be enabled for TensorFlow
Do you wish to build TensorFlow with CUDA support? [y/N]
No CUDA support will be enabled for TensorFlow
......
INFO: Starting clean (this may take a while). Consider using --async if the clean takes more than several minutes.
```

And, the following build commands:
```
bazel build --config=opt //tensorflow/tools/pip_package:build_pip_package
bazel-bin/tensorflow/tools/pip_package/build_pip_package /tmp/tensorflow_pkg
sudo pip install /tmp/tensorflow_pkg/tensorflow-1.2.0rc1-cp27-cp27mu-linux_x86_64.whl
```

**Comment by michaelisard** [comment]
We aren't currently supporting the AMDGPU backend unfortunately, so marking 'community support'.

**Comment by adityaatluri** [comment]
Hi @michaelisard ,
I am try to get llvm.BUILD file get right. The code isn't the issue

**Comment by petecoup** [comment]
_ZN4llvm6AMDGPU9isComputeEj == llvm::AMDGPU::isCompute(unsigned int)
This looks like it is in lib/Target/AMDGPU/Utils.
So, looks like you need to build a amdgpu_utils library, like aarch64_utils.  Then, add it as a dependency on the required libraries.

**Comment by adityaatluri** [comment]
It took me 5 hours to figure it out (well, 20 minutes of build time for each run). I added 
```
"@llvm//:amdgpu_code_gen",
"@llvm//::amdgpu_utils",
"@llvm//:amdgpu_disassembler",
```
to ./tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/BUILD (I first tried with cpu/BUILD and it worked).
Build still running will let you know if it messes up.

**Comment by adityaatluri** [comment]
Hi,
I am able to pass build stage. But the library `_pywrap_tensorflow_internal.so` do not contain `LLVMInitializeAMDGPU` variants. Here is `nm -S` dump for different targets

```
/usr/local/lib/python2.7/dist-packages/tensorflow/python$ nm -S _pywrap_tensorflow_internal.so | grep LLVMInitializeAArch64
00000000035caed0 0000000000000032 t LLVMInitializeAArch64AsmPrinter
00000000036915c0 0000000000000059 t LLVMInitializeAArch64Disassembler
0000000003686580 00000000000000b8 t LLVMInitializeAArch64Target
00000000036ecc70 000000000000007d t LLVMInitializeAArch64TargetInfo
00000000036ab9b0 0000000000000113 t LLVMInitializeAArch64TargetMC
00000000036eca30 0000000000000008 t _ZZ31LLVMInitializeAArch64TargetInfoENUlN4llvm6Triple8ArchTypeEE_4_FUNES1_
```
For AMDGPU,
```
/usr/local/lib/python2.7/dist-packages/tensorflow/python$ nm -S _pywrap_tensorflow_internal.so | grep LLVMInitializeAMDGPU
/usr/local/lib/python2.7/dist-packages/tensorflow/python$
```

The files in `./external/llvm/lib/Target/AMDGPU` do contain the function declarations
```
~/tensorflow/bazel-tensorflow/external/llvm/lib/Target/AMDGPU$ grep -r LLVMInitialize  *

AMDGPUAsmPrinter.cpp:extern "C" void LLVMInitializeAMDGPUAsmPrinter() {
AMDGPUTargetMachine.cpp:extern "C" void LLVMInitializeAMDGPUTarget() {
AsmParser/AMDGPUAsmParser.cpp:extern "C" void LLVMInitializeAMDGPUAsmParser() {
Disassembler/AMDGPUDisassembler.cpp:extern "C" void LLVMInitializeAMDGPUDisassembler() {
MCTargetDesc/AMDGPUMCTargetDesc.cpp:extern "C" void LLVMInitializeAMDGPUTargetMC() {
TargetInfo/AMDGPUTargetInfo.cpp:extern "C" void LLVMInitializeAMDGPUTargetInfo() {
```

**Comment by adityaatluri** [comment]
Hi @petecoup 
I forked tf and made my changes to llvm-amdgpu branch.
https://github.com/adityaatluri/tensorflow/tree/llvm-amdgpu

**Comment by petecoup** [comment]
Yes.  For the CPU targets, in compiler/xla/service/cpu/cpu_compiler.cc, there is CpuCompiler::InitializeLLVMTarget, where CPU targets can add their own set of these initialization routines.  Looks like the NVPTX backend is initialized in compiler/xla/servce/gpu/llvm_gpu_backend/gpu_backend_lib.cc.  Don't know what that means for where to put AMDGPU...good luck!

**Comment by adityaatluri** [comment]
Hi @petecoup ,
I am not concerned with code right now. If I could get LLVM to build AMDGPU target inside TF, that would make life lot easier. Is it possible for you to try out my forked version of TF? As long as the tf shared library has all the routines inside AMDGPU target, I can figure out a way to use them inside xla code.

**Comment by annanay25** [comment]
@adityaatluri, these targets are defined in the [BUILD](https://github.com/tensorflow/tensorflow/blob/master/third_party/llvm/llvm.BUILD#L25-L31) file. You could try adding AMDGPU to this list. You will also have to take care of any ```.def``` files that LLVM uses for this backend though.

**Comment by adityaatluri** [comment]
@annanay25 I already did https://github.com/adityaatluri/tensorflow/blob/llvm-amdgpu/third_party/llvm/llvm.BUILD#L826

**Comment by annanay25** [comment]
Okay, then I am in a very similar position as you are right now!

Ref: https://github.com/tensorflow/tensorflow/issues/10471

TLDR;
I build Polly with LLVM, but I am unable to get Polly's passes initialized in the XLA pipeline.
A solution to your issue will hopefully give some insight to mine as well :)

**Comment by petecoup** [comment]
Not sure what you mean "...not concerned with code right now".  When compiling the offline compiler, tfcompile, directly calling these LLVMInitializeAMDGPU* routines is the only way you can get the target initialized so you can generate code for it (and I don't think it will even link in the guts of your target otherwise).  I imagine it is the same for the python wrapper.  For CPU targets, this is where I mentioned earlier, compiler/xla/service/cpu/cpu_compiler.cc:CpuCompiler::InitializeLLVMTarget.

**Comment by adityaatluri** [comment]
What I meant to say is, once I get AMDGPU target to build, I can reference it inside xla code. Having xla code supporting the target and not able to link/build is unproductive use of time. I have gone through xla code and looks like a straight forward implementation for AMDGPU backend. I need help with getting AMDGPU target built right.

**Comment by adityaatluri** [comment]
@michaelisard 
Can you assign this to some one at google who can help with this issue. Currently this is a major hurdle in enabling AMDGPU backend for xla. Once I can see LLVMInitializeAMDGPU* symbols inside _pywrap_tensorflow_internal.so, I am good to go! 
Thanks!

**Comment by michaelisard** [comment]
@hawkinsp kindly volunteered to help out.

**Comment by adityaatluri** [comment]
@michaelisard @hawkinsp Thank you very much. Appreciate the effort! 👍

**Comment by hawkinsp** [comment]
@adityaatluri I have a pending change that adds bazel build rules for the AMDGPU LLVM backend that should land in the next few days. That takes care of building the backend.

Once that it is landed, you will need to do at least these three things to use the backend:
a) uncomment `"AMDGPU"` in the `llvm_targets` list at the top of the build file.
It will appear roughly here once the change lands:
https://github.com/tensorflow/tensorflow/blob/master/third_party/llvm/llvm.BUILD#L25
I left it commented out for now to avoid making builds longer since there are no actual users yet.

b) add a bazel build dependency on `"@llvm//:amdgpu_target"` from somewhere.
c) add a call to `LLVMInitializeAMDGPUTarget()`.

Note that (c) is essential --- if a symbol is not used then bazel will strip it from the binary. That's probably what's going wrong with your attempt.

As an experiment I added such a call here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/cpu/cpu_compiler.cc#L127
and I verified that LLVMInitializeAMDGPUTarget was present in the pywrap_tensorflow.so symbols. (Obviously the XLA *CPU* plugin isn't the right place in reality --- you really want to add it to the GPU plugin or add a separate AMDGPU plugin, depending on how similar you expect the two to be.)

**Comment by adityaatluri** [comment]
@hawkinsp Awesome!
Thank you for the insight into the bazel build process. I was able to do first 2 steps. I'll check for step 3.
Do you know a time frame where the patch will be upstreamed?
Do you have any plans to extending it to Stream Executor?

**Comment by hawkinsp** [comment]
The patch will be upstreamed in the next few days, whenever we sync up our branch with the Github repository.

I'm afraid I don't have time to work on a AMDGPU Tensorflow StreamExecutor plugin, nor access to the necessary hardware. But if you are interested in working on one, we can definitely give you pointers.

One thought: XLA actually only uses a relatively small subset of the StreamExecutor API, so it isn't quite as imposing as it might seem. I believe the only essential pieces are those for things like opening the device, allocating memory, and launching kernels.

**Comment by adityaatluri** [comment]
Thank you for offering help, do you expect to change XLA architecture to fit in AMDGPU code gen along with NVPTX? Because, the triple is hard coded to nvptx [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/gpu_compiler.cc#L81)

**Comment by hawkinsp** [comment]
I don't think we have any very specific thoughts on how this should best be done.

There is a somewhat under-publicized mailing list for folks working on XLA development which might make a better forum for this sort of wide-ranging discussion than a Github issue:
https://groups.google.com/forum/#!forum/xla-dev

To answer your specific question: I suspect the right thing to do is to have a separate AMDGPU XLA plugin, separate from the NVidia GPU XLA plugin. The two can share much or even most of the code (as do the current GPU and CPU backends, which share most of their LLVM logic, for example), but details like the target triple, or exactly which compiler passes to use, can differ between plugins. 

As I mentioned, we don't really have a lot of time ourselves to work on an additional backend, but we would be very happy if one did exist, so this is a "contributions welcome" situation.

**Comment by adityaatluri** [comment]
Hi @hawkinsp 
Thank you!
